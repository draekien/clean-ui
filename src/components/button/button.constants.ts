import { alpha } from '@theme-ui/color';
import { Theme } from '@theme-ui/css';
import { ButtonColorDefinitions } from './button.types';

/** The colors used in the primary button variant. */
export const primaryColors: ButtonColorDefinitions = {
  normal: {
    content: 'b-000',
    border: 'p-400',
    background: 'p-400',
  },
  hover: {
    border: 'p-300',
    background: 'p-300',
  },
  active: {
    border: 'p-200',
    background: 'p-200',
  },
};

/** The colors used in the secondary button variant. */
export const secondaryColors: ButtonColorDefinitions = {
  normal: {
    content: 'b-000',
    border: 's-400',
    background: 's-400',
  },
  hover: {
    border: 's-300',
    background: 's-300',
  },
  active: {
    border: 's-200',
    background: 's-200',
  },
};

/** The colors used in the outline button variant */
export const outlineColors: ButtonColorDefinitions = {
  normal: {
    content: 'p-400',
    border: 'p-400',
    background: 'transparent',
  },
  hover: {
    border: 'p-300',
    background: 'transparent',
  },
  active: {
    border: 'p-200',
    background: 'transparent',
  },
};

/** The colors used in the text and link button variants. */
export const textColors: ButtonColorDefinitions = {
  normal: {
    content: 'p-400',
    border: 'p-400',
  },
  hover: {
    content: 'p-300',
    border: 'p-300',
    background: (t: Theme) => `${alpha('b-200', 0.25)(t)}`,
  },
  active: {
    content: 'p-200',
    border: 'p-200',
    background: (t: Theme) => `${alpha('b-300', 0.25)(t)}`,
  },
};

/** The colors used in the gradient button variant. */
export const gradientColors: ButtonColorDefinitions = {
  normal: {
    content: 'b-000',
    border: 'transparent',
    background: (t: Theme) =>
      'linear-gradient(to right, ' +
      alpha('p-400', 1)(t) +
      ' 0%, ' +
      alpha('s-400', 1)(t) +
      ' 51%, ' +
      alpha('p-400', 1)(t) +
      ' 100%)',
  },
};
