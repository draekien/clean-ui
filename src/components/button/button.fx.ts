import { useState, useEffect } from 'react';

export const useDynamicIcon = (icon?: string) => {
  const [state, setState] = useState(icon);

  useEffect(() => {
    setState(icon);
  }, [icon]);

  return { icon: state };
};
