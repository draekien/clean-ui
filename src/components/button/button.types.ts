import { Color, ComputedColor, Spacing } from '../../types/styles';

/** The two different positions the addons can be rendered at. */
export type ButtonAddonPosition = 'left' | 'right';

/** The different button variants that can be rendered. */
export type ButtonVariant =
  | 'primary'
  | 'secondary'
  | 'outline'
  | 'gradient'
  | 'text'
  | 'link';

/** A button color definition. */
export interface ColorDefinitions {
  /** The content color. */
  content?: ComputedColor | Color;
  /** The border color. */
  border?: ComputedColor | Color;
  /** The background color. */
  background?: ComputedColor | Color;
}

/** The colors of the button. */
export interface ButtonColorDefinitions {
  /** The colors for the default state of the button. */
  normal: ColorDefinitions;
  /** The colors for the hovered state of the button. */
  hover?: ColorDefinitions;
  /** The colors for the active state of the button. */
  active?: ColorDefinitions;
}

/** The padding of the button. */
export interface ButtonPadding {
  /** The left padding. */
  pl?: Spacing;
  /** The right padding. */
  pr?: Spacing;
  /** The horizontal padding. */
  px?: Spacing;
  /** The vertical padding. */
  py?: Spacing;
}
