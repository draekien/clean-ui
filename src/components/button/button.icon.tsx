/** @jsxImportSource theme-ui */
import { FC } from 'react';
import { iconWrapperCss } from './button.styles';
import { ButtonAddonPosition } from './button.types';
import { Size } from '../../types/layouts';
import { Icon } from '../icon/icon';

/** The props required to render the button's icon. */
export interface ButtonIconProps {
  /**
   * The material icon name.
   * @remarks
   * No icon will be rendered if a name is not provided.
   */
  name?: string;
  /**
   * The size of the icon to render.
   * @remarks
   * See {@link Size} for more details.
   *
   * @defaultValue `'small'`
   */
  size?: Size;
  /**
   * The position of the icon on the button.
   * @remarks
   * See {@link ButtonAddonPosition} for more details.
   */
  position: ButtonAddonPosition;
  /**
   * The requested render position of the icon on the button.
   * The icon will only be rendered when this value matches the
   * `position` property.
   * @remarks
   * See {@link ButtonAddonPosition} for more details.
   */
  renderPosition?: ButtonAddonPosition;
  /**
   * Is the button in the loading state?
   * @remarks
   * The icon will not be rendered if loading is `true`
   */
  loading?: boolean;
  /** Does the button have any children? */
  buttonHasChildren?: boolean;
}

/**
 * Renders a button icon.
 * @param props - the {@link ButtonIconProps}.
 * @returns A button icon component
 */
export const ButtonIcon: FC<ButtonIconProps> = ({
  name,
  position,
  renderPosition,
  size = 'small',
  loading,
  buttonHasChildren,
}) => {
  if (!name) return null;
  if (loading) return null;
  if (renderPosition !== position) return null;

  return (
    <span sx={iconWrapperCss(renderPosition, !!buttonHasChildren)}>
      <Icon name={name} size={size} />
    </span>
  );
};
