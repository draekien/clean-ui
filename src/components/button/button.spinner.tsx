/** @jsxImportSource theme-ui */
import { FC } from 'react';
import { Size } from '../../types/layouts';
import { Spinner, SpinnerVariant } from '../spinner/spinner';
import { iconWrapperCss } from './button.styles';
import { ButtonAddonPosition } from './button.types';

/** The props required to render the button's spinner. */
export interface ButtonSpinnerProps {
  /**
   * The spinner variant to render.
   * @remarks
   * See {@link SpinnerVariant} for more details.
   * @defaultValue `'spiral'`
   */
  variant?: SpinnerVariant;
  /**
   * The size of the spinner to render.
   * @remarks
   * See {@link Size} for more details.
   * @defaultValue `'small'`
   */
  size?: Size;
  /**
   * Is the button in a loading state?
   * @remarks
   * The spinner will only render if loading is `true`
   */
  loading?: boolean;
  /**
   * The position of the spinner on the button.
   * @remarks
   * See {@link ButtonAddonPosition} for more details.
   */
  position?: ButtonAddonPosition;
  /**
   * The requested render position of the spinner on the button.
   * The spinner will only be rendered when this value matches the `position` property.
   * @remarks
   * See {@link ButtonAddonPosition} for more details.
   */
  renderPosition?: ButtonAddonPosition;
  /**
   * Does the button show content when loading?
   */
  showContentWhileLoading?: boolean;
}

/**
 * Renders a button's spinner.
 * @param props - the {@link ButtonSpinnerProps}
 * @returns A button spinner component.
 */
export const ButtonSpinner: FC<ButtonSpinnerProps> = ({
  variant = 'spiral',
  size = 'small',
  loading,
  position,
  renderPosition,
  showContentWhileLoading,
}) => {
  if (!loading) return null;
  if (renderPosition !== position) return null;
  if (!renderPosition) return null;

  return (
    <span sx={iconWrapperCss(renderPosition, !!showContentWhileLoading)}>
      <Spinner variant={variant} size={size} />
    </span>
  );
};
