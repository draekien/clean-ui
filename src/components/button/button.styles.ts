import { ButtonSize, ButtonVariant } from './button';
import { alpha } from '@theme-ui/color';
import { Theme, ThemeUIStyleObject } from '@theme-ui/css';

export interface ButtonCssProps {
  variant: ButtonVariant;
  size: ButtonSize;
  disabled: boolean;
  loading: boolean;
  fullWidth: boolean;
  active: boolean;
  feature: boolean;
  showContentWhileLoading: boolean;
  iconPosition?: 'left' | 'right';
  hasIcon?: boolean;
  hasText?: boolean;
  circle?: boolean;
}

const getColors = (variant: ButtonVariant) => {
  if (variant === 'secondary') {
    return {
      backgroundColor: 's-400',
      borderColor: 's-400',
      color: 'b-000',
      hoverBackground: 's-300',
      hoverBorder: 's-300',
      activeBackground: 's-200',
      activeBorder: 's-200',
    };
  } else if (variant === 'text') {
    return {
      color: 'p-400',
      hoverColor: 'p-300',
      activeColor: 'p-200',
      hoverBackground: (t: Theme) => `${alpha('b-200', 0.25)(t)}`,
      activeBackground: (t: Theme) => `${alpha('b-300', 0.25)(t)}`,
    };
  } else if (variant === 'link') {
    return {
      color: 'p-400',
      hoverColor: 'p-300',
      activeColor: 'p-200',
      hoverBackground: (t: Theme) => `${alpha('b-200', 0.25)(t)}`,
      activeBackground: (t: Theme) => `${alpha('b-300', 0.25)(t)}`,
    };
  } else if (variant === 'outline') {
    return {
      backgroundColor: 'transparent',
      borderColor: 'p-400',
      color: 'p-400',
      hoverBackground: 'transparent',
      hoverBorder: 'p-300',
      activeBackground: 'transparent',
      activeBorder: 'p-200',
    };
  } else if (variant === 'gradient') {
    return {
      backgroundColor: (t: Theme) =>
        `linear-gradient(to right, ${alpha('p-400', 1)(t)} 0%, ${alpha(
          's-400',
          1
        )(t)} 51%, ${alpha('p-400', 1)(t)} 100%)`,
      borderColor: 'transparent',
      color: 'b-000',
    };
  }

  // default to primary
  return {
    backgroundColor: 'p-400',
    borderColor: 'p-400',
    color: 'b-000',
    hoverBackground: 'p-300',
    hoverBorder: 'p-300',
    activeBackground: 'p-200',
    activeBorder: 'p-200',
  };
};

export const buttonCss = (props: ButtonCssProps): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    border: '1px solid',
    borderRadius: 'sm',
    boxShadow: 'md',
    cursor: 'pointer',
    variant: 'text.body',
    textAlign: 'center',
    textTransform: 'none',
    transition: 'all 300ms',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    textDecoration: 'none',
  };

  if (props.fullWidth) {
    css.width = '100%';
    css.display = 'flex';

    if (props.variant === 'text' || props.variant === 'link') {
      css.justifyContent =
        props.iconPosition && props.hasIcon ? 'space-between' : 'flex-start';
    }
  }

  if (props.feature) {
    props.variant = 'gradient';
    props.size = 'large';

    css.textTransform = 'uppercase';
    css.letterSpacing = '0.5rem';
  }

  const colors = getColors(props.variant);

  if (props.hasText) {
    if (props.size === 'small') {
      css.px = 'md';
      css.py = 'xxs';
    } else if (props.size === 'large') {
      css.py = 'xs';
      css.px = 'xxl';
    } else {
      css.py = 'xxs';
      css.px = 'xl';
    }
  } else {
    if (props.size === 'small') {
      css.p = 'xxs';
    } else if (props.size === 'large') {
      css.p = 'sm';
    } else {
      css.p = 'xs';
    }

    if (props.circle) {
      css.borderRadius = 'max';
    }
  }

  if ((!props.loading || props.showContentWhileLoading) && props.hasText) {
    if (props.iconPosition === 'left') {
      switch (props.size) {
        case 'small':
          css.pl = 'xs';
          css.pr = 'sm';
          break;
        case 'medium':
          css.pl = 'sm';
          css.pr = 'md';
          break;
        default:
          css.pl = 'md';
          css.pr = 'lg';
          break;
      }
    } else if (props.iconPosition === 'right') {
      switch (props.size) {
        case 'small':
          css.pl = 'sm';
          css.pr = 'xs';
          break;
        case 'medium':
          css.pl = 'md';
          css.pr = 'sm';
          break;
        default:
          css.pl = 'lg';
          css.pr = 'md';
          break;
      }
    }
  }

  if (props.variant === 'gradient') {
    css.backgroundSize = '200% auto';
    css.border = '1px solid transparent';
    css.position = 'relative';
    css[':before'] = {
      content: '""',
      position: 'absolute',
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      zIndex: -1,
      margin: -1,
      borderRadius: 'inherit',
      backgroundImage: colors.backgroundColor,
    };
  }

  if (props.variant === 'text') {
    css.background = 'transparent';
    css.border = 'none';
    css.boxShadow = 'none';
    css.px = 'xxs';
  }

  if (props.variant === 'link') {
    css.background = 'transparent';
    css.border = 'none';
    css.boxShadow = 'none';
    css.position = 'relative';
    css.pl = 'xxs';
    css.pr = 'xxs';
    css[':before'] = {
      content: '""',
      position: 'absolute',
      backgroundColor: 'primary',
      height: 2,
      left: 0,
      right: 0,
      bottom: 0,
      transition: 'all 300ms',
    };
  }

  if (!props.disabled) {
    if (props.variant === 'outline') {
      css.backgroundColor = colors.backgroundColor;
      css.color = colors.borderColor;
      css.span = {
        color: colors.borderColor,
        textShadow: 'text',
        transition: 'all 300ms',
      };
      css[':hover, :focus'] = {
        color: colors.hoverBorder,
        span: {
          color: colors.hoverBorder,
        },
      };
      css[':active'] = {
        color: colors.activeBorder,
        span: {
          color: colors.activeBorder,
        },
      };
    } else if (props.variant === 'gradient') {
      css.backgroundImage = colors.backgroundColor;
      css.color = colors.color;
      css[':hover, :focus'] = {
        backgroundPosition: 'right center',
      };
      css[':active'] = {
        backgroundPosition: 'center',
      };
    } else if (props.variant === 'text') {
      css.color = colors.color;
      css[':hover, :focus'] = {
        backgroundColor: colors.hoverBackground,
        color: colors.hoverColor,
      };
      css[':active'] = {
        backgroundColor: colors.activeBackground,
        color: colors.activeColor,
      };
    } else if (props.variant === 'link') {
      css.color = colors.color;
      css[':hover, :focus'] = {
        backgroundColor: colors.hoverBackground,
        [':before']: {
          backgroundColor: colors.hoverColor,
        },
      };
      css[':active'] = {
        backgroundColor: colors.activeBackground,
        [':before']: {
          backgroundColor: colors.activeColor,
        },
      };
    } else {
      css.backgroundColor = colors.backgroundColor;
      css.borderColor = colors.borderColor;
      css.color = colors.color;
      css.span = {
        color: colors.color,
      };
      css[':hover, :focus'] = {
        backgroundColor: colors.hoverBackground,
        borderColor: colors.hoverBorder,
      };
      css[':active'] = {
        backgroundColor: colors.activeBackground,
        borderColor: colors.activeBorder,
      };
    }

    if (
      props.active &&
      props.variant !== 'gradient' &&
      props.variant !== 'text' &&
      props.variant !== 'link'
    ) {
      css.backgroundColor = colors.activeBackground;
      css.borderColor = colors.activeBackground;
    } else if (props.active && props.variant === 'gradient') {
      css.backgroundPosition = 'center';
    }
  } else {
    css.cursor = 'not-allowed';
    css.boxShadow = 'disabled';
    if (props.variant === 'gradient') {
      css.backgroundImage = (t: Theme) =>
        `linear-gradient(to right, ${alpha('p-000', 1)(t)} 0%, ${alpha(
          's-000',
          1
        )(t)} 51%, ${alpha('p-000', 1)(t)} 100%)`;
      css.color = 'b-000';
    } else if (props.variant === 'text') {
      css.boxShadow = 'none';
      css.color = 'b-300';
    } else if (props.variant === 'link') {
      css.color = 'b-300';
      css.boxShadow = 'none';
      css[':before'] = {
        content: '""',
        position: 'absolute',
        backgroundColor: 'b-300',
        height: 2,
        left: 0,
        right: 0,
        bottom: 0,
        transition: 'all 300ms',
      };
    } else {
      css.backgroundColor = 'muted';
      css.borderColor = 'muted';
      css.color = 'b-300';
    }
  }

  return css;
};

export const iconWrapperCss = (
  position: 'left' | 'right',
  hasText: boolean
): ThemeUIStyleObject => {
  const css: any = {
    display: 'inline-flex',
    textDecoration: 'none',
  };

  if (hasText) {
    if (position === 'left') {
      css.marginRight = 'xs';
    } else {
      css.marginLeft = 'xs';
    }
  } else {
    css['i'] = {
      fontSize: '1.5rem',
    };
  }

  return css;
};
