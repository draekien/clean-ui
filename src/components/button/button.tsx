// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React from 'react';
import * as styles from './button.styles';
import { useDynamicIcon } from './button.fx';
import { Spinner, SpinnerVariant } from '../spinner/spinner';
import { Icon } from '../icon/icon';
import { Size } from '../../types/layouts';

export type ButtonVariant =
  | 'primary'
  | 'secondary'
  | 'outline'
  | 'gradient'
  | 'text'
  | 'link';
export type ButtonSize = 'small' | 'medium' | 'large';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** the type of button to render
   * @defaultValue 'primary'
   */
  variant?: ButtonVariant;
  /** the size of the rendered button
   * @defaultValue 'medium'
   */
  size?: ButtonSize;
  /** the name of the material icon to render */
  icon?: string;
  iconSize?: Size;
  /** the position of the icon in the button */
  iconPosition?: 'left' | 'right';
  /** should the button render as disabled?
   * @defaultValue false
   */
  disabled?: boolean;
  /** should the button take up the full width of the parent container?
   * @defaultValue false
   */
  fullWidth?: boolean;
  /** should the button render as loading?
   * @defaultValue false
   */
  loading?: boolean;
  /** should the button render content while in the loading state?
   * @defaultValue false
   */
  showContentWhileLoading?: boolean;
  /** should the button render in the active state?
   * @defaultValue false
   */
  active?: boolean;
  /** should the button render in the featured state?
   * @defaultValue false
   */
  feature?: boolean;
  /** the spinner variant to use when loading
   * @defaultValue 'spiral'
   */
  spinnerVariant?: SpinnerVariant;
  /** should the button be rendered as a circle?
   *  this only works when there is no button text.
   *  @defaultValue 'false'
   */
  circle?: boolean;
  onClick?: (e: React.SyntheticEvent<HTMLButtonElement, MouseEvent>) => any;
}

export const Button: React.FC<ButtonProps> = ({
  variant = 'primary',
  size = 'medium',
  disabled = false,
  fullWidth = false,
  loading = false,
  active = false,
  feature = false,
  showContentWhileLoading = false,
  iconSize = 'small',
  iconPosition,
  children,
  spinnerVariant = 'spiral',
  circle = false,
  ...rest
}: ButtonProps) => {
  const { icon } = useDynamicIcon(rest.icon);

  const renderSpacedIcon = (renderPosition: 'left' | 'right') => {
    if (!icon) return null;
    if (loading) return null;
    if (renderPosition !== iconPosition) return null;

    return (
      <span sx={styles.iconWrapperCss(renderPosition, !!children)}>
        <Icon name={icon} size={iconSize} />
      </span>
    );
  };

  const renderSpacedSpinner = (renderPosition: 'left' | 'right') => {
    if (!loading) return null;
    if (renderPosition !== iconPosition && iconPosition) return null;
    if (!iconPosition && renderPosition === 'left') return null;

    return (
      <span sx={styles.iconWrapperCss(renderPosition, !!showContentWhileLoading)}>
        <Spinner variant={spinnerVariant} size={iconSize} />
      </span>
    );
  };

  const isDisabledOrLoading = disabled || loading;

  return (
    <button
      sx={styles.buttonCss({
        variant,
        size,
        disabled: isDisabledOrLoading,
        fullWidth,
        loading,
        active,
        feature,
        showContentWhileLoading,
        iconPosition,
        hasText: !!children,
        hasIcon: !!icon,
        circle,
      })}
      disabled={isDisabledOrLoading}
      {...rest}>
      {renderSpacedSpinner('left')}
      {renderSpacedIcon('left')}
      {(!loading || showContentWhileLoading) && children}
      {renderSpacedIcon('right')}
      {renderSpacedSpinner('right')}
    </button>
  );
};
