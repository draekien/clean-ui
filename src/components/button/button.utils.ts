import {
  primaryColors,
  secondaryColors,
  outlineColors,
  textColors,
  gradientColors,
} from './button.constants';
import { ButtonVariant, ButtonColorDefinitions } from './button.types';

/**
 * Gets the color definitions for the specified button variant.
 * @param variant - the {@link ButtonVariant}
 * @returns The {@link ButtonColorDefinitions} for the specified variant
 */
export const getButtonColors = (variant: ButtonVariant): ButtonColorDefinitions => {
  switch (variant) {
    case 'secondary':
      return secondaryColors;
    case 'outline':
      return outlineColors;
    case 'text':
    case 'link':
      return textColors;
    case 'gradient':
      return gradientColors;
    default:
      return primaryColors;
  }
};
