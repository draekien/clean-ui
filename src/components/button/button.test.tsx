import React from 'react';
import { hexToRgb, themedRender } from '../../helpers/testHelpers';
import { colors } from '../theme/theme.colors';
import { layouts } from '../theme/theme.layouts';
import { Button } from './button';

describe('Button Component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(<Button>Button</Button>);

    expect(container).toMatchSnapshot();
  });

  describe('Button Variants', () => {
    it('Should render with correct styles when variant is primary', () => {
      const { container } = themedRender(<Button>Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.backgroundColor).toBe(hexToRgb(colors['p-400']));
      expect(styles.borderColor).toBe(colors['p-400'].toLowerCase());
      expect(styles.color).toBe(hexToRgb(colors['b-000']));
    });

    it('Should render with correct styles when variant is secondary', () => {
      const { container } = themedRender(<Button variant="secondary">Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.backgroundColor).toBe(hexToRgb(colors['s-400']));
      expect(styles.borderColor).toBe(colors['s-400'].toLowerCase());
      expect(styles.color).toBe(hexToRgb(colors['b-000']));
    });

    it('Should render with correct styles when variant is outline', () => {
      const { container } = themedRender(<Button variant="outline">Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.backgroundColor).toBe('rgba(0, 0, 0, 0)');
      expect(styles.color).toBe(hexToRgb(colors['p-400']));
    });

    it('Should render with correct styles when variant is gradient', () => {
      const { container } = themedRender(<Button variant="gradient">Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.backgroundSize).toBe('200% auto');
      expect(styles.color).toBe(hexToRgb(colors['b-000']));
    });

    it('Should render with correct styles when variant is gradient and button is disabled', () => {
      const { container } = themedRender(
        <Button variant="gradient" disabled>
          Button
        </Button>
      );

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.backgroundSize).toBe('200% auto');
      expect(styles.color).toBe(hexToRgb(colors['b-000']));
    });

    it('Should render with centered background for active gradient button', () => {
      const { container } = themedRender(
        <Button variant="gradient" active>
          Button
        </Button>
      );

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.backgroundPosition).toBe('center');
    });
  });

  describe('Button flags', () => {
    it('Should render with correct styles when disabled prop is true', () => {
      const { container } = themedRender(<Button disabled>Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.cursor).toBe('not-allowed');
      expect(styles.backgroundColor).toBe(hexToRgb(colors['b-200']));
      expect(styles.color).toBe(hexToRgb(colors['b-300']));
    });

    it('Should render with correct styles when fullWidth prop is true', () => {
      const { container } = themedRender(<Button fullWidth>Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('flex');
    });

    it('Should render with correct styles when feature prop is true', () => {
      const { container } = themedRender(<Button feature>Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.display).toBe('inline-flex');
      expect(styles.textTransform).toBe('uppercase');
      expect(styles.letterSpacing).toBe('0.5rem');
      expect(styles.backgroundSize).toBe('200% auto');
    });

    it('Should render with correct styles when active prop is true', () => {
      const { container } = themedRender(<Button active>Button</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.backgroundColor).toBe(hexToRgb(colors['p-200']));
    });

    it('Should render as a circle when circle prop is true and button is icon only', () => {
      const { container } = themedRender(<Button icon="done" circle />);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.borderRadius).toBe('999px');
    });

    it('Should render with the correct styles when variant is a link', () => {
      const { container } = themedRender(<Button variant="link">Link</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.background).toBe('transparent');
      expect(styles.color).toBe(hexToRgb(colors.primary));
    });

    it('Should render with the correct styles when variant is a text', () => {
      const { container } = themedRender(<Button variant="text">Text</Button>);

      const styles = window.getComputedStyle(
        container.querySelector('button') as Element
      );
      expect(styles.background).toBe('transparent');
      expect(styles.color).toBe(hexToRgb(colors.primary));
    });
  });

  describe('Button Sizes', () => {
    describe('Icon only', () => {
      it('Should render with correct styles with size is small', () => {
        const { container } = themedRender(<Button size="small" icon="done" />);

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );
        expect(styles.padding).toBe('0.25rem');
      });

      it('Should render with correct styles with size is medium', () => {
        const { container } = themedRender(<Button size="medium" icon="done" />);

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );
        expect(styles.padding).toBe('0.5rem');
      });

      it('Should render with correct styles with size is large', () => {
        const { container } = themedRender(<Button size="large" icon="done" />);

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );
        expect(styles.padding).toBe('0.75rem');
      });
    });

    describe('Has text', () => {
      it('Should render with correct styles with size is small', () => {
        const { container } = themedRender(<Button size="small">Button</Button>);

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );
        expect(styles.paddingLeft).toBe(layouts.space.md);
        expect(styles.paddingRight).toBe(layouts.space.md);
      });

      it('Should render with correct styles with size is medium', () => {
        const { container } = themedRender(<Button size="medium">Button</Button>);

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );
        expect(styles.paddingLeft).toBe(layouts.space.xl);
        expect(styles.paddingRight).toBe(layouts.space.xl);
      });

      it('Should render with correct styles with size is large', () => {
        const { container } = themedRender(<Button size="large">Button</Button>);

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );
        expect(styles.paddingLeft).toBe(layouts.space.xxl);
        expect(styles.paddingRight).toBe(layouts.space.xxl);
      });
    });
  });

  describe('Button with icon', () => {
    ['left' as const, 'right' as const].forEach((pos) => {
      it(`Should render small button correctly for icon ${pos}`, () => {
        const { container } = themedRender(
          <Button size="small" icon="done" iconPosition={pos}>
            Button
          </Button>
        );

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );

        if (pos === 'left') {
          expect(styles.paddingLeft).toBe(layouts.space.xs);
          expect(styles.paddingRight).toBe(layouts.space.sm);
        } else {
          expect(styles.paddingLeft).toBe(layouts.space.sm);
          expect(styles.paddingRight).toBe(layouts.space.xs);
        }
      });

      it(`Should render medium button correctly for icon ${pos}`, () => {
        const { container } = themedRender(
          <Button size="medium" icon="done" iconPosition={pos}>
            Button
          </Button>
        );

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );

        if (pos === 'left') {
          expect(styles.paddingLeft).toBe(layouts.space.sm);
          expect(styles.paddingRight).toBe(layouts.space.md);
        } else {
          expect(styles.paddingLeft).toBe(layouts.space.md);
          expect(styles.paddingRight).toBe(layouts.space.sm);
        }
      });

      it(`Should render large button correctly for icon ${pos}`, () => {
        const { container } = themedRender(
          <Button size="large" icon="done" iconPosition={pos}>
            Button
          </Button>
        );

        const styles = window.getComputedStyle(
          container.querySelector('button') as Element
        );

        if (pos === 'left') {
          expect(styles.paddingLeft).toBe(layouts.space.md);
          expect(styles.paddingRight).toBe(layouts.space.lg);
        } else {
          expect(styles.paddingLeft).toBe(layouts.space.lg);
          expect(styles.paddingRight).toBe(layouts.space.md);
        }
      });
    });
  });
});
