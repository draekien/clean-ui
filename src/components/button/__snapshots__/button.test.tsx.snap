// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Should match snapshot 1`] = `
<div>
  <button
    class="css-1cf1hkw-Button"
  >
    Button
  </button>
</div>
`;
